Schema Evolution Phase I: LinkDef refactor

tomeichlersmith opened the pull request:
I am updating _ldmx-sw_, here are the details.

### What are the issues that this addresses?
This is the first (of two) phases of #1815 where I refactor the event dictionary and CMake infrastructure:
- no more double-pass over the module CMakeLists.txt, that allows a pretty drastic simplification but means **we have to specify the event libraries that we depend on** (`SimCore::Event`) instead of inheriting from them being linked to framework.
  - This specification is particularly apparent for "stand-alone" processors as described in https://ldmx-software.github.io/using/analysis/ldmx-sw.html where users need to add `needs = 'Ecal_Event'` to their python config.
- I needed to refactor how GENIE was being configured in the CMake to get this to work. This was my whole day today (sorry @wesketchum for my grumpy messages!) but I persevered!
- While refactoring CMake, I also did some other cleanup work that I hope makes the CMake infrastructure easier to understand in the future.
  - updated how the Geant4 target was made, trying to make the `BuildMacros.cmake` file less all-encompassing.
  - I dropped support for Reco-only or Sim-only builds. I don't think folks use this much and it is mainly leftover from a time before the container images. It would be a pain to maintain since many of the "Reco" modules still depend on the simulation module via the event objects.
  - moved executables into `app` directories so they are not compiled into the library and as an executable
  - fix-up documentation comments at top of `cmake/Find` scripts


## Check List
- [x] I successfully compiled _ldmx-sw_ with my developments.
- [x] I read, understood and follow the [coding rules](https://ldmx-software.github.io/developing/coding-rules.html).
- [x] Compiling, linking, and running basic simulation
- [x] ~Have a `GENIE::Reweight` and `GENIE::Generator` targets in CMake? They are two different projects...~ Not doing this because we don't use them separated. Our `GenieReweightProducer` uses stuff from both Reweight and Generator, so it needs both anyways. Separating them provides no help to us at this time.
- [x] Run validation samples from development build
- [x] Double check a production image build works. I mocked this by `mv build hide-build` so that it appeared like the build wasn't there (since it is not copied into the production image).
- [x] Remove `link off` lines in LinkDef files that are ignored in our version of ROOT (and future versions)
- [x] Update documentation of "stand-alone" processors https://ldmx-software.github.io/using/analysis/ldmx-sw.html

tomeichlersmith committed:
WIP move LinkDef to committed-file, refactor CMake

- compiles and runs a basic simulation
- does not link the run_test complaining about missing GENIE libraries

tomeichlersmith committed:
tidy up central cmake, can't partition build anymore

tomeichlersmith committed:
add necessary dependencies to some genie libs

tomeichlersmith committed:
attempt to refactor, strugglign to get GENIE to link properly

tomeichlersmith committed:
fallback to passing list of libraries to link to

- not working still, linking issues when attempting to load anything
  linked to GENIE

tomeichlersmith committed:
try to include dependencies

tomeichlersmith committed:
copy over updated set of dependencies within GENIE

tomeichlersmith committed:
undo movement into Geo, keep movement of g4-vis

tomeichlersmith committed:
use GENIE::GENIE target instead of variables

tomeichlersmith committed:
ignore wget history file

github-actions[bot] committed:
Apply clang-format

tomeichlersmith committed:
necessary links to event libraries

tomeichlersmith committed:
ignore intermediate genie data files

tomeichlersmith committed:
remove 'link off' lines in LinkDef

these lines are ignored by ROOT 6+ and so they are just clutter

tomeichlersmith committed:
don't need to install the HepMC3 root dictionary

its already in a linker-findable path so ROOT is able to find it at
runtime

tomeichlersmith committed:
remove duplicate VERSION project specifiers

the project calls in module CMakeLists.txt are just there to define the
module's name

tomeichlersmith committed:
remove debug messages and alt config option in CMake

tomeichlersmith commented on cmake/FindGENIE.cmake:
```diff
@@ -0,0 +1,107 @@
+  return()
+endif()
+
+set(GENIE_INCLUDE_DIRS "/usr/local/include/GENIE")
```
Do we want to update this to not include the `GENIE` subdirectory? This would then force all tine `#include` statements to have `GENIE` in them when using `GENIE` stuff (which might be helpful?)

tomeichlersmith replied:
While this is "Outdated" because I replaced the hard-coded `/usr/local` with the variable `GENIE_DIR` defined as `/usr/local` for consistency with other CMake Config scripts, my question remains. Looking for insight from @wesketchum or another developer.

tomeichlersmith replied:
Actually nevermind, I tried this and GENIE's own headers use the include path without the `GENIE` prefix, so we need it in the include directories anyways. I'm not interested in updating the CMake config to try to make that be private, so I'm leaving this as-is.

tomeichlersmith committed:
move executables into an app subdirectory

tomeichlersmith committed:
make sure all paths use the same CMake variables

just in case things change

tomeichlersmith committed:
cleanup comments in cmake/Find scripts

bryngemark commented on cmake/FindGENIE.cmake:
```diff
@@ -0,0 +1,108 @@
+# Find an installation of GENIE. Highly tuned to the installation
+# of GENIE that is built into the ldmx/dev image.
+#
+# This will define the following variables
```
and GENIE_DIR? picky documentation comment :D

bryngemark reviewed (approved):
this is at the level that if you tell me this works, I believe you. out of curiosity, what was the deal with the versions in all the CMakeLists projects?

tomeichlersmith committed:
update documentation comment on cmake/FindGENIE

tomeichlersmith committed:
load dependency modules for stand-alone processors

I can't get the linking to behave as intended, so I'm doing the hammer
approach and just manually including any dependency libraries as
libraries to be loaded by the Process at runtime

tvami reviewed (approved):
Tests pass, the PNet scores are little different, but not so that the KS would flag it
